shared/tls/sni/chi.go

```diff
@@ -0,0 +1,84 @@
+package sni
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"io"
+	"net"
+	"time"
+)
+
+// Conn is a net.Conn with custom Reader
+type Conn struct {
+	net.Conn
+	io.Reader
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	return c.Reader.Read(b)
+}
+
+// PeekClientHelloInfo extracts the ClientHelloInfo from a connection
+// still allowing a future handler have complete untouched access to
+// the stream
+func PeekClientHelloInfo(ctx context.Context, conn net.Conn) (
+	*tls.ClientHelloInfo, net.Conn, error) {
```

Comment:
maybe we should try this pattern when breaking long function definitions:
```go
func PeekClientHelloInfo(ctx context.Context, conn net.Conn) (
     *tls.CientHelloInfo, net.Conn, error,
) {
```

Reply:
@karasz what about now?

Reply:
+1